File: Ch1/C++/ZeroMatrix.cpp
#include <stdlib.h>

#include <ctime>
#include <iostream>
#include <string>
#include <unordered_set>
#include <vector>

void print2DVector(const std::vector<std::vector<int>>& matrix) {
    for (std::vector<int> row : matrix) {
        std::cout << "[";
        for (int i = 0; i < row.size(); i++) {
            std::cout << row.at(i) << (i != row.size() - 1 ? ", " : "");
        }
        std::cout << "]" << std::endl;
    }
    std::cout << std::endl;
}

void zeroMatrix(std::vector<std::vector<int>>& matrix) {
    std::unordered_set<int> rowSet, colSet;
    for (int row = 0; row < matrix.size(); row++) {
        for (int col = 0; col < matrix[row].size(); col++) {
            if (matrix[row][col] == 0) {
                rowSet.emplace(row);
                colSet.emplace(col);
            }
        }
    }
    for (int row = 0; row < matrix.size(); row++) {
        for (int col = 0; col < matrix[row].size(); col++) {
            if (rowSet.count(row) > 0 || colSet.count(col) > 0) {
                matrix[row][col] = 0;
            }
        }
    }
}

int main() {
    srand(time(NULL));
    int numRows = rand() % 10 + 1, numCols = rand() % 10 + 1;
    std::vector<std::vector<int>> matrix(numRows, std::vector<int>(numCols, 0));
    for (int i = 0; i < numRows; i++) {
        for (int j = 0; j < numCols; j++) {
            matrix[i][j] = rand() % 10;
        }
    }
    print2DVector(matrix);
    zeroMatrix(matrix);
    print2DVector(matrix);
    return 0;
}

File: Ch1/Java/ZeroMatrix.java
import java.util.HashSet;
import java.util.Arrays;
public class ZeroMatrix {

    public static void main(String[] args) {
        int numRows = (int) (Math.random()*10) + 1;
        int numCols = (int) (Math.random()*10) + 1;
        int[][] matrix = new int[numRows][numCols];
        for (int i = 0; i < numRows; i++) {
            for (int j = 0; j < numCols; j++) {
                matrix[i][j] = (int) (Math.random()*10);
            }
        }

        print2DArray(matrix);
        zeroMatrix(matrix);
        print2DArray(matrix);
    }

    public static void print2DArray(int[][] arr) {
        System.out.printf("%s\n\n", Arrays.deepToString(arr).replace("], ", "]\n"));
    }

    public static void zeroMatrix(int[][] matrix) {
        HashSet<Integer> rowSet = new HashSet<Integer>();
        HashSet<Integer> colSet = new HashSet<Integer>();
        for (int row = 0; row < matrix.length; row++) {
            for (int col = 0; col < matrix[row].length; col++) {
                if (matrix[row][col] == 0) {
                    rowSet.add(row);
                    colSet.add(col);
                }
            }
        }
        for (int row = 0; row < matrix.length; row++) {
            for (int col = 0; col < matrix[row].length; col++) {
                if (rowSet.contains(row) || colSet.contains(col)) {
                    matrix[row][col] = 0;
                }
            }
        }
    }



}